src/components/views/rooms/DecryptionFailureBar.tsx

```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { useCallback, useContext, useEffect, useState } from 'react';
+import { Room } from 'matrix-js-sdk/src/models/room';
+import { MatrixEvent } from 'matrix-js-sdk/src/models/event';
+
+import Modal from '../../../Modal';
+import { _t } from '../../../languageHandler';
+import defaultDispatcher from "../../../dispatcher/dispatcher";
+import { Action } from "../../../dispatcher/actions";
+import AccessibleButton from '../elements/AccessibleButton';
+import { OpenToTabPayload } from "../../../dispatcher/payloads/OpenToTabPayload";
+import { UserTab } from "../dialogs/UserTab";
+import MatrixClientContext from "../../../contexts/MatrixClientContext";
+import SetupEncryptionDialog from '../dialogs/security/SetupEncryptionDialog';
+import { SetupEncryptionStore } from '../../../stores/SetupEncryptionStore';
+import Spinner from "../elements/Spinner";
+
+interface IProps {
+    room: Room;
+    failures: MatrixEvent[];
+}
+
+// Number of milliseconds to display a loading spinner before prompting the user for action
+const WAIT_PERIOD = 5000;
+
+export const DecryptionFailureBar: React.FC<IProps> = ({ failures, room }) => {
+    const context = useContext(MatrixClientContext);
+
+    // Display a spinner for a few seconds before presenting an error message,
+    // in case keys are about to arrive
+    const [waiting, setWaiting] = useState<boolean>(true);
+    useEffect(() => {
+        const timeout = setTimeout(() => setWaiting(false), WAIT_PERIOD);
+        return () => clearTimeout(timeout);
+    }, []);
+
+    // Is this device unverified?
+    const [needsVerification, setNeedsVerification] = useState<boolean>(false);
+    // Does this user have verified devices other than this device?
+    const [hasOtherVerifiedDevices, setHasOtherVerifiedDevices] = useState<boolean>(false);
+    // Does this user have key backups?
+    const [hasKeyBackup, setHasKeyBackup] = useState<boolean>(false);
+
+    // Keep track of session IDs that do not have pending key requests
```

Comment:
could you be clear (somewhere) what exactly a "pending key request" means? In other words, when is a key request no longer considered to be pending? (When the message is decrypted? If we get a "rejected" response? Never?)

Reply:
Also: given that the js-sdk automatically sends a key request for any event it fails to decrypt, what could cause this list to be non-empty?

Reply:
As I understand it, a request will remain in the list of outgoing key requests until it's cancelled or until we decrypt all messages from that session. You're right that in theory, this list should always be empty. However, here's the situation:
- I know, from personal experience, that the "Request keys from other devices" link has sometimes been genuinely helpful in resolving problems. Given that, I don't want to remove it entirely.
- You've raised valid concerns about the performance impact of repeatedly cancelling and resending key requests, and you don't want this link/button to cancel and resend in-flight requests.
- Given that, I'd at least like to have this around in the unexpected situation where a key request *isn't* in flight for an undecrypted message. Ideally, we should absolutely understand why, if ever, that would happen, and make sure it doesn't. However, I don't know how long it will take us to get to that ideal situation, and in the meantime, I want users to have as many options as possible for resolving the immediate error.

Reply:
> As I understand it, a request will remain in the list of outgoing key requests until it's cancelled or until we decrypt all messages from that session. 

Please could you stick this in a comment (if we end up keeping it, per the discussion below):

> I know, from personal experience, that the "Request keys from other devices" link has sometimes been genuinely helpful in resolving problems. Given that, I don't want to remove it entirely.

yeees, but that *is* a "cancel and resend" button, so I don't think that's a fair comparison.

> Given that, I'd at least like to have this around in the unexpected situation where a key request isn't in flight for an undecrypted message.

OK, but on the flip side, this means we have a bunch of code and CSS which we can't test outside a unit test, and which we never expect the user to see (if they do see it, they have found a bug). Generally I feel like the eleweb codebase is baroque enough as it is without making it even more complicated!

At the very very least we need to comment this to say that `unrequestedSessions` will only be non-empty in the event of a js-sdk bug.

Reply:
Actually, I'm wrong there - there are absolutely situations where we'll receive an undecryptable message and won't send key requests for it. In fact, one such situation does occur and is specifically tested in the Cypress tests on this PR. If we know that a session was specifically withheld, then we don't bother sending a request for it; however, verifying our device (as advised by the call to action) might mean that a new key share request will be accepted.

Reply:
In any case, here's my proposal: we show a "Request keys" button even if all of our messages have outgoing requests, and it cancels and resends those; *however*, it will only do this once per message (within the lifetime of the component), and only when specifically requested by a user. In other words, it will keep track of every session it has already sent or re-sent a key request for, and won't re-send requests for those sessions; the button won't be displayed unless there's some session visible that we haven't already requested via that button.

Reply:
>  If we know that a session was specifically withheld, then we don't bother sending a request for it; however, verifying our device (as advised by the call to action) might mean that a new key share request will be accepted.

yup, fair point. Could you put a comment in citing this example?

> In any case, here's my proposal: we show a "Request keys" button even if all of our messages have outgoing requests, and it cancels and resends those; however, it will only do this once per message (within the lifetime of the component), and only when specifically requested by a user.

I'm still a bit worried that we're basically training users to smash the button whenever anything goes wrong, but it seems like a reasonable proposal. Let's give it a go.

Reply:
Done.